### PR TITLE
Add a chat plugin for Mafia

### PR DESCRIPTION
**Pull request by bumbadadabum**
This is still very unpolished and incomplete, but it'd be very nice to be able to test this more on the main server.

### CODE REVIEW

**Comment by Zarel** [comment]
Yeah, we're putting these images on the client

File: chat-plugins/mafia.js [hunk 1]
```diff
@@ -0,0 +1,610 @@
...
+
+	announcementWindow(image, content) {
+		this.room.add('|html|' + this.mafiaWindow(image, content));
+		this.room.update();
+	}
+
+	updateVotes() {
+		let text = '';
+		for (let i in this.currentVote) {
+			text += this.players[i].name + ': ' + this.currentVote[i] + ' votes.<br/>';
```
Comment (Slayer95) [inline chat-plugins/mafia.js]: Escaping missing

File: chat-plugins/mafia.js [hunk 2]
```diff
@@ -0,0 +1,610 @@
...
+		output += '<br/><p>Who do you wish to target?</p>';
+		for (let i in this.validTargets) {
+			output += '<button value="/mafia target ' + this.validTargets[i].userid + '" name="send">' + this.validTargets[i].name + '</button>';
+		}
+
+		this.sendRoom('|uhtml|mafia' + this.game.room.gameNumber + 'target' + this.game.gamestate + this.game.day + '|' + this.game.mafiaWindow(image, output));
+	}
+
+	updateTarget(image) {
+		this.sendRoom('|uhtmlchange|mafia' + this.game.room.gameNumber + 'target' + this.game.gamestate + this.game.day + '|' + this.game.mafiaWindow(image, 'Targeting ' + this.target.name + '!'));
```
Comment (Slayer95) [inline chat-plugins/mafia.js]: Escape missing

File: chat-plugins/mafia.js [hunk 3]
```diff
@@ -0,0 +1,610 @@
...
+	}
+
+	voteWindow(image, content) {
+		let output = content;
+		output += '<br/><p>Who do you wish to vote for?</p>';
+		for (let i in this.validVotes) {
+			output += '<button value="/mafia vote ' + this.validVotes[i].userid + '" name="send">' + this.validVotes[i].name + '</button>';
+		}
+
+		console.log("Showing shit to " + this.name);
```
Comment (Slayer95) [inline chat-plugins/mafia.js]: (ಠ_ಠ)

File: chat-plugins/mafia.js [hunk 4]
```diff
@@ -0,0 +1,610 @@
...
+
+	updateTarget(image) {
+		this.sendRoom('|uhtmlchange|mafia' + this.game.room.gameNumber + 'target' + this.game.gamestate + this.game.day + '|' + this.game.mafiaWindow(image, 'Targeting ' + this.target.name + '!'));
+	}
+
+	voteWindow(image, content) {
+		let output = content;
+		output += '<br/><p>Who do you wish to vote for?</p>';
+		for (let i in this.validVotes) {
+			output += '<button value="/mafia vote ' + this.validVotes[i].userid + '" name="send">' + this.validVotes[i].name + '</button>';
```
Comment (Slayer95) [inline chat-plugins/mafia.js]: Escape missing here as well

File: chat-plugins/mafia.js [hunk 5]
```diff
@@ -0,0 +1,610 @@
...
+			if ((side === 'all' || thisSide === side || (side === 'nomafia' && thisSide !== 'mafia')) && (targetSelf || this.game.players[i] !== this)) {
+				this.validTargets[i] = this.game.players[i];
+			}
+		}
+	}
+
+	// Triggers after the user has selected their target.
+	onReceiveTarget(target) {
+		if (!this.targeting) {
+			return;
```
Comment (Slayer95) [inline chat-plugins/mafia.js]: This is a silent failure for  a direct command

File: chat-plugins/mafia.js [hunk 6]
```diff
@@ -0,0 +1,610 @@
...
+			this.updateTarget(this.class.image);
+
+			for (let i in this.game.players) {
+				if (this.game.players[i].voting || this.game.players[i].targeting) {
+					return;
+				}
+			}
+			this.game.progress();
+		} else {
+			this.sendRoom("Invalid target");
```
Comment (Slayer95) [inline chat-plugins/mafia.js]: Early return would be more readable probably

File: chat-plugins/mafia.js [hunk 7]
```diff
@@ -0,0 +1,610 @@
...
+			this.game.updateVotes();
+
+			for (let i in this.game.players) {
+				if (this.game.players[i].voting || this.game.players[i].targeting) {
+					return;
+				}
+			}
+			this.game.progress();
+		} else {
+			this.sendRoom("You can't vote for that person");
```
Comment (Slayer95) [inline chat-plugins/mafia.js]: Ditto

File: chat-plugins/mafia.js [hunk 8]
```diff
@@ -0,0 +1,610 @@
...
+			this.toExecute = this.class[event].function;
+		}
+
+		this.game.executionOrder.push(this);
+	}
+
+	kill(message) {
+		if (this.invincible) return;
+
+		this.game.announcementWindow(deadImage, message + '<br/>' + this.name + ', the ' + this.class.name + ' lies dead on the ground.');
```
Comment (Slayer95) [inline chat-plugins/mafia.js]: Escape the name

File: chat-plugins/mafia-data.js [hunk 9]
```diff
@@ -0,0 +1,99 @@
+// Data for the mafia chat plugin.
+
+'use strict';
+
+// This object contains all functions that execute in the callback function of any mafia class. Executed from the context of the executing player.
+// Target is a MafiaPlayer object.
+let MafiaFunctions = {
+	copReport: function (target) {
+		if (target.class.side === 'village') {
+			return 'After investigating ' + target.name + ' you find out they\'re sided with the village.';
```
Comment (Slayer95) [inline chat-plugins/mafia-data.js]: Convention is to use double quotes for human-readable text.
Then you no longer need to escape the single quotes.

File: chat-plugins/mafia.js [hunk 10]
```diff
@@ -0,0 +1,610 @@
...
+					player.roleBlocked = false;
+					player.toExecute = null;
+				} else {
+					let output;
+					if (player.target) {
+						output = player.toExecute(player.target);
+					} else {
+						output = player.toExecute();
+					}
+
```
Comment (Slayer95) [inline chat-plugins/mafia.js]: This output needs to be escaped as it includes player names.

File: chat-plugins/mafia-data.js [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+		flavorText: 'You are the hooker. Every night, you can visit someone in the town. The person you visit can\'t execute any actions that night.',
+
+		onNight: {
+			target: {side: 'all', type: 'single'},
```
Comment (Slayer95) [inline chat-plugins/mafia-data.js]: Suggestions: ``any`` instead of ``all``. ``count`` instead of ``type``.

File: chat-plugins/mafia.js [hunk 12]
```diff
@@ -0,0 +1,610 @@
...
+		this.gameid = 'hangman';
+		this.title = 'Mafia';
+		this.allowRenames = false;
+		this.playerCap = max;
+		this.PlayerClass = MafiaPlayer;
+
+		this.roles = roles;
+		this.day = 1;
+		this.gamestate = 'pregame';
+		this.timeout = null;
```
Comment (Slayer95) [inline chat-plugins/mafia.js]: Did you mean ``timer``?

File: chat-plugins/mafia.js [hunk 13]
```diff
@@ -0,0 +1,610 @@
...
+			if (room.game) return this.errorReply("There is already a game in progress in this room.");
+
+			// TODO: make a generator for a default setup.
+			if (!params) return this.errorReply("No roles entered.");
+
+			for (let i = 0; i < params.length; i++) {
+				if (!(params[i] in MafiaData.MafiaClasses)) {
+					return this.errorReply(Tools.escapeHTML(params[i]) + " is not a valid mafia class.");
+				}
+			}
```
Comment (Slayer95) [inline chat-plugins/mafia.js]: This seems to be a very cumbersome syntax

``/mafia new town, town, town, town, mafia, mafia,...``

It should probably be something like ``/mafia new 10 town, 8 mafia, 1 doctor``, etc

File: chat-plugins/mafia-data.js (hunk 9 above)
Comment (Zarel) [inline chat-plugins/mafia-data.js]: But then you have to escape the double quotes!

There's just no winning. :(

File: chat-plugins/mafia.js (hunk 3 above)
Comment (bumbadadabum) [inline chat-plugins/mafia.js]: oops left a debug message in

**Comment by bumbadadabum** [comment]
I tried to fix the issues listed.

**Comment by bumbadadabum** [comment]
Also added the images from the client

File: chat-plugins/mafia.js [hunk 14]
```diff
@@ -0,0 +1,609 @@
...
+			delete room.mafiaEnabled;
+			if (room.chatRoomData) {
+				delete room.chatRoomData.mafiaEnabled;
+				Rooms.global.writeChatRoomData();
+			}
+			return this.sendReply("Mafia has been disabled for this room.");
+		},
+
+		enable: function (target, room, user) {
+			if (!this.can('tournamentsmanagement', null, room)) return;
```
Comment (Zarel) [inline chat-plugins/mafia.js]: Let's give it its own permission; I don't want random people to test it at this point.

Comment (Zarel) [inline chat-plugins/mafia.js]: I can merge it after this change.

**Comment by bumbadadabum** [comment]
Ok I did that

**Pull request merged**